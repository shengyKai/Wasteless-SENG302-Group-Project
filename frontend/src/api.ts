/*
 * Created on Wed Feb 10 2021
 *
 * The Unlicense
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or distribute
 * this software, either in source code form or as a compiled binary, for any
 * purpose, commercial or non-commercial, and by any means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors of this
 * software dedicate any and all copyright interest in the software to the public
 * domain. We make this dedication for the benefit of the public at large and to
 * the detriment of our heirs and successors. We intend this dedication to be an
 * overt act of relinquishment in perpetuity of all present and future rights to
 * this software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

/**
 * Declare all available services here
 */
import axios from 'axios';
const SERVER_URL = process.env.VUE_APP_SERVER_ADD;

const instance = axios.create({
  baseURL: SERVER_URL,
  timeout: 2000,
  withCredentials: true,
});

type MaybeError<T> = T | string;

export type User = {
  id: number,
  firstName: string,
  lastName: string,
  middleName?: string,
  nickname?: string,
  bio?: string,
  email: string,
  dateOfBirth: string,
  phoneNumber?: string,
  homeAddress: Location,
  created?: string,
  role?: "user" | "globalApplicationAdmin" | "defaultGlobalApplicationAdmin",
  businessesAdministered?: Business[],
};

export type Location = {
  streetNumber?: string,
  streetName?: string,
  city?: string,
  region?: string,
  country: string,
  postcode?: string,
};

export type CreateUser = {
  firstName: string,
  lastName: string,
  middleName?: string,
  nickname?: string,
  bio?: string,
  email: string,
  dateOfBirth: string,
  phoneNumber?: string,
  homeAddress: Location,
  password: string,
};

export type BusinessType = 'Accommodation and Food Services' | 'Retail Trade' | 'Charitable organisation' | 'Non-profit organisation';

export const BUSINESS_TYPES: BusinessType[] = ['Accommodation and Food Services', 'Retail Trade', 'Charitable organisation', 'Non-profit organisation'];

export type Business = {
  id: number,
  administrators?: User[],
  name: string,
  description?: string,
  address: string,
  businessType: BusinessType,
  created?: string,
};

export type CreateBusiness = {
  name: string,
  description?: string,
  address: string,
  businessType: BusinessType,
};

export type Image = {
  id: number,
  filename: string,
  thumbnailFilename: string,
};

export type Product = {
  id: string,
  name: string,
  description?: string,
  manufacturer?: string,
  recommendedRetailPrice?: number,
  created?: string,
  images: Image[],
};

export type CreateProduct = Omit<Product, 'created' | 'images'>;

function isLocation(obj: any): obj is Location {
  if (obj === null || typeof obj !== 'object') return false;
  if (obj.streetNumber !== undefined && typeof obj.streetNumber !== 'string') return false;
  if (obj.streetName !== undefined && typeof obj.streetName !== 'string') return false;
  if (obj.city !== undefined && typeof obj.city !== 'string') return false;
  if (obj.region !== undefined && typeof obj.region !== 'string') return false;
  if (typeof obj.country !== 'string') return false;
  if (obj.postcode !== undefined && typeof obj.postcode !== 'string') return false;

  return true;
}

function isUser(obj: any): obj is User {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.id !== 'number') return false;
  if (typeof obj.firstName !== 'string') return false;
  if (typeof obj.lastName !== 'string') return false;
  if (obj.middleName !== undefined && typeof obj.middleName !== 'string') return false;
  if (obj.nickname !== undefined && typeof obj.nickname !== 'string') return false;
  if (obj.bio !== undefined && typeof obj.bio !== 'string') return false;
  if (typeof obj.email !== 'string') return false;
  if (typeof obj.dateOfBirth !== 'string') return false;
  if (obj.phoneNumber !== undefined && typeof obj.phoneNumber !== 'string') return false;
  if (!isLocation(obj.homeAddress)) return false;
  if (obj.created !== undefined && typeof obj.created !== 'string') return false;
  if (obj.role !== undefined && !['user', 'globalApplicationAdmin', 'defaultGlobalApplicationAdmin'].includes(obj.role))
    return false;
  if (obj.businessesAdministered !== undefined && !isBusinessArray(obj.businessesAdministered)) return false;

  return true;
}

function isBusiness(obj: any): obj is Business {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.id !== 'number') return false;
  if (obj.administrators !== undefined && !isUserArray(obj.administrators)) return false;
  if (typeof obj.name !== 'string') return false;
  if (obj.description !== undefined && typeof obj.description !== 'string') return false;
  if (typeof obj.address !== 'object') return false;
  if (!BUSINESS_TYPES.includes(obj.businessType)) return false;
  if (obj.created !== undefined && typeof obj.created !== 'string') return false;

  return true;
}

function isNumberArray(obj: any): obj is number[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (typeof elem !== 'number') return false;
  }
  return true;
}

function isUserArray(obj: any): obj is User[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isUser(elem)) return false;
  }
  return true;
}

function isBusinessArray(obj: any): obj is Business[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isBusiness(elem)) return false;
  }
  return true;
}

type OrderBy = 'userId' | 'relevance' | 'firstName' | 'middleName' | 'lastName' | 'nickname' | 'email' | 'address';

/**
 * Sends a search query to the backend.
 *
 * @param query Query string to search for
 * @param pageIndex Index of page to start the results from (1 = first page)
 * @param resultsPerPage Number of results to return per page
 * @param orderBy Specifies the method used to sort the results
 * @param reverse Specifies whether to reverse the search results (default order is descending for relevance and ascending for all other orders)
 * @returns List of user infos for the current page or an error message
 */
export async function search(query: string, pageIndex: number, resultsPerPage: number, orderBy: OrderBy, reverse: boolean): Promise<MaybeError<User[]>> {
  let response;
  try {
    response = await instance.get('/users/search', {
      params: {
        searchQuery: query,
        page: pageIndex,
        resultsPerPage,
        orderBy,
        reverse: reverse.toString(),
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (!isUserArray(response.data)) {
    return 'Response is not user array';
  }

  return response.data;
}

/**
 * Sends a query for the number of search results for a given query to the backend.
 *
 * @param query Query string to search for
 * @returns Number of search results or an error message
 */
export async function getSearchCount(query: string): Promise<MaybeError<number>> {
  let response;
  try {
    response = await instance.get('/users/search/count', {
      params: {
        searchQuery: query,
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (typeof response.data?.count !== 'number') {
    return 'Response is not number';
  }

  return response.data.count;
}

/**
 * Queries the backend for a specific user by their id.
 *
 * @param id User id, if ommitted then fetch the logged in user's info
 * @returns User info for the given id or an error message
 */
export async function getUser(id: number): Promise<MaybeError<User>> {
  let response;
  try {
    response = await instance.get('/users/' + id);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (!isUser(response.data)) {
    return 'Response is not user';
  }

  return response.data;
}

/**
 * Logs in to the given user account by setting the authentication cookie.
 *
 * @param email User email
 * @param password User password
 * @returns The now logged in user ID if operation is successful, otherwise a string error.
 */
export async function login(email: string, password: string): Promise<MaybeError<number>> {
  let response;
  try {
    response = await instance.post('/login', {
      email: email,
      password: password,
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid credentials';
    return `Request failed: ${status}`;
  }
  let id = response.data.userId;
  if (typeof id !== 'number') return 'Invalid response';

  return id;
}

/**
 * Creates a user with the given properties.
 * Note: This doesn't automatically log the user in.
 *
 * @param user Initial user properties
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createUser(user: CreateUser): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/users', user);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid create user request';
    if (status === 409) return 'Email in use';
    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Makes the given user a GAA, if the current user is a DGAA.
 *
 * @param userId User to make GAA
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function makeAdmin(userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.post(`/users/${userId}/makeAdmin`);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';
    if (status === 406) return 'User does not exist';

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Revokes the given user's admin rights
 *
 * @param userId User to revoke permissions
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function revokeAdmin(userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.post(`/users/${userId}/revokeAdmin`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';
    if (status === 406) return 'User does not exist';

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Creates a business
 *
 * @param business The properties to create the business with
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createBusiness(business: CreateBusiness): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/businesses', business);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';

    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Add a product to a businesses catalogue.
 *
 * @param businessId The business to add the product to
 * @param product The properties to create a product with
 * @return undefined if operation is successful, otherwise a string error
 */
export async function createProduct(businessId: number, product: CreateProduct): Promise<MaybeError<undefined>> {
  try {
    await  instance.post(`/businesses/${businessId}/products`, product);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';

    if (status === 400) {
      // TODO Not sure exactly how the backend is going to communicate with us that the product code
      // is unavailable.

      // eslint-disable-next-line no-constant-condition
      if (false) {
        return 'Product code unavailable';
      }

      return 'Invalid parameters';
    }

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Fetches a business with the given id.
 *
 * @param businessId Business id to fetch
 * @returns The requested business if operation is successful, otherwise a string error.
 */
export async function getBusiness(businessId: number): Promise<MaybeError<Business>> {
  let response;
  try {
    response = await instance.get(`/businesses/${businessId}`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 406) return 'Business not found';

    return 'Request failed: ' + status;
  }

  if (!isBusiness(response.data)) {
    return 'Invalid response type';
  }

  return response.data;
}