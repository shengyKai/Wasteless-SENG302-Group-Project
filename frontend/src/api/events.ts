import {Keyword, MarketplaceCard, MarketplaceCardSection, MaybeError, Message, Sale, User} from "./internal";
import axios from 'axios';
import { is } from 'typescript-is';

const SERVER_URL = process.env.VUE_APP_SERVER_ADD;

const instance = axios.create({
  baseURL: SERVER_URL,
  timeout: 5 * 1000,
  withCredentials: true,
});

export type AnyEvent = GlobalMessageEvent | ExpiryEvent | DeleteEvent | KeywordCreatedEvent | MessageEvent | InterestEvent;

export type EventTag = 'none' | 'red' | 'orange' | 'yellow' | 'green' | 'blue' | 'purple'
export type EventStatus = 'normal' | 'starred' | 'archived'

type BaseEvent<T extends string> = {
  id: number,
  created: string,
  type: T,
  tag: EventTag,
  status: EventStatus,
  read: boolean,
  lastModified: string
}

export type GlobalMessageEvent = BaseEvent<'GlobalMessageEvent'> & {
  message: string
}

export type ExpiryEvent = BaseEvent<'ExpiryEvent'> & {
  card: MarketplaceCard
}

export type DeleteEvent = BaseEvent<'DeleteEvent'> & {
  title: string,
  section: MarketplaceCardSection
}

export type KeywordCreatedEvent = BaseEvent<'KeywordCreatedEvent'> & {
  keyword: Keyword,
  creator: User
}
export type MessageEvent = BaseEvent<'MessageEvent'> & {
  message: Message,
  participantType: 'buyer' | 'seller',
  conversation: {
    buyer: User,
    card: MarketplaceCard,
    id: number,
  },
}
export type InterestEvent = BaseEvent<'InterestEvent'> & {
  saleItem: Sale,
  interested: boolean,
}

export type PurchaseEvent = BaseEvent<'PurchaseEvent'> & {
  saleItem: Sale
}

/**
 * Updates an event as read. No body is needed in this case as the backend would only have to change
 * a boolean value, and subsequently, the event emitters will retrieve the events with the updated field.
 * @param eventId Event id of the event to mark as read
 */
export async function updateEventAsRead(eventId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.put(`/feed/${eventId}/read`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return "Invalid authorization for marking this event";
    if (status === 406) return 'Event does not exist';
    return 'Request failed: ' + error.response?.data.message;
  }
  return undefined;
}

/**
 * Updates an event to a selected status with overwriting strategy. Overwriting the event's status
 * Archive is a one way action, no UI component available for overwriting archived event
 * @param eventId Event id of the event to be overwriting
 * @param status  The desire status to be updated for the event
 */
export async function updateEventStatus(eventId: number, status: EventStatus): Promise<MaybeError<undefined>> {
  try {
    await instance.put(`/feed/${eventId}/status`, {
      value: status
    }
    );
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return "Invalid authorization for modifying event status";
    if (status === 406) return 'Event does not exist';
    return 'Request failed: ' + error.response?.data.message;
  }
  return undefined;
}
/**
 * Retrieve events for the newsfeed of the user with the given id. If a modifedSince timestamp is provided,
 * only events that have been after this timestamp will be retrieved. If no timestamp is provided, all events
 * associated with the user will be retrieved.
 * @param userId The ID number of the user to retrieve newsfeed events for.
 * @param modifiedSince A datetime string in UTC format. If this string is provided, only events modified after
 * this timestamp will be retrieved.
 * @returns An array of events to be displayed in the user's newsfeed.
 */
export async function getEvents(userId: number, modifiedSince: string | undefined): Promise<MaybeError<AnyEvent[]>> {
  const params = new URLSearchParams();
  if (modifiedSince) {
    params.append("modifiedSince", modifiedSince);
  }
  try {
    const response = await instance.get(`/users/${userId}/feed`, {
      params: params
    });
    if (!is<AnyEvent[]>(response.data)) {
      return 'Response is not an event array';
    }
    return response.data;
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid \'modified since\' date';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return `Invalid authorisation for getting events associated with user ${userId}`;
    if (status === 406) return 'User does not exist';
    return 'Request failed: ' + error.response?.data.message;
  }
}