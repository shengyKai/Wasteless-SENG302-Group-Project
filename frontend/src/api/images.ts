import axios from 'axios';
import { is } from 'typescript-is';
import { MaybeError, Image } from './internal';

const SERVER_URL = process.env.VUE_APP_SERVER_ADD;

const instance = axios.create({
  baseURL: SERVER_URL,
  timeout: 5 * 1000,
  withCredentials: true,
});

/**
 * Upload an image to the application. Response will contain the location of the uploaded image and its associated id.
 * @param image The expected image file
 */
export async function uploadImage(image: File): Promise<MaybeError<Image>> {
  try {
    let formData = new FormData();
    formData.append('file', image);
    const response = await instance.post(`/media/images`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    });
    if (!is<Image>(response.data)) {
      return 'Image was not recieved';
    }
    return response.data;
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid image: ' + error.response?.data.message;
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return 'Operation not permitted';
    if (status === 413) return 'Image too large';
    return 'Request failed: ' + error.response?.data.message;
  }
}