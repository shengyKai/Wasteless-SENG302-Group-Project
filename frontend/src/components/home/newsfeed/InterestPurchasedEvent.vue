<template>
  <Event :event="event" title="An item you have liked has been purchased" :error="errorMessage">
    <v-card-text class="pb-1">
      <strong>{{itemBought}}</strong>
      from
      <strong>
        <router-link
          class="text--secondary"
          :to="`/business/${business.id}`"
        >
          {{business.name}}
        </router-link>
      </strong>
      has been sold.
    </v-card-text>
  </Event>
</template>

<script>
import Event from "@/components/home/newsfeed/Event";

export default {
  name: "InterestPurchasedEvent",
  components: {
    Event
  },
  props: {
    event: {
      type: Object,
      required: true,
    },
  },
  computed: {
    itemBought() {
      return this.boughtSaleItem.quantity + "x " + this.boughtSaleItem.product.name;
    },
    boughtSaleItem() {
      return this.event.boughtSaleItem;
    },
    product() {
      return this.boughtSaleItem.product;
    },
    business() {
      return this.product.business;
    },
  },
};
</script>

<style scoped>

</style>