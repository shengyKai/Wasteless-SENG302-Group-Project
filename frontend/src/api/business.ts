import {is} from 'typescript-is';
import {Location, MaybeError, SearchResults, instance} from "@/api/internal";
import {User} from "@/api/user";
import { Image } from '@/api/images';

export type BusinessType = 'Accommodation and Food Services' | 'Retail Trade' | 'Charitable organisation' | 'Non-profit organisation';
export const BUSINESS_TYPES: BusinessType[] = ['Accommodation and Food Services', 'Retail Trade', 'Charitable organisation', 'Non-profit organisation'];
type BusinessOrderBy = 'created' | 'name' | 'location' | 'businessType';

export type Business = {
  id: number,
  primaryAdministratorId: number,
  administrators?: User[],
  name: string,
  description?: string,
  address: Location,
  businessType: BusinessType,
  created?: string,
  images?: Image[],
};

export type CreateBusiness = {
  primaryAdministratorId: number,
  name: string,
  description?: string,
  address: Location,
  businessType: BusinessType,
};

export type ModifyBusiness = {
  primaryAdministratorId: number,
  name: string,
  description?: string,
  address: Location,
  businessType: BusinessType,
  updateProductCountry: boolean
}

/**
 * Creates a business
 *
 * @param business The properties to create the business with
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createBusiness(business: CreateBusiness): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/businesses', business);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'You have been logged out. Please login again and retry';

    return error.response.data.message;
  }

  return undefined;
}

/**
 * Fetches a business with the given id.
 *
 * @param businessId Business id to fetch
 * @returns The requested business if operation is successful, otherwise a string error.
 */
export async function getBusiness(businessId: number): Promise<MaybeError<Business>> {
  let response;
  try {
    response = await instance.get(`/businesses/${businessId}`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 406) return 'Business not found';

    return 'Request failed: ' + status;
  }

  if (!is<Business>(response.data)) {
    return 'Invalid response type';
  }

  return response.data;
}

/**
 * Sends a business search query to the backend. Can specify query, business type or both.
 *
 * @param query Query string to search for
 * @param businessType Business type to filter search results by
 * @param pageIndex Index of page to start the results from (1 = first page)
 * @param resultsPerPage Number of results to return per page
 * @param orderBy Specifies the method used to sort the results
 * @param reverse Specifies whether to reverse the search results (default order is descending for relevance and ascending for all other orders)
 * @returns List of business infos for the current page or an error message
 */
export async function searchBusinesses(query: string | undefined, businessType: BusinessType | undefined, pageIndex: number, resultsPerPage: number, orderBy: BusinessOrderBy, reverse: boolean): Promise<MaybeError<SearchResults<Business>>> {
  let response;
  try {
    response = await instance.get('/businesses/search', {
      params: {
        searchQuery: query,
        businessType: businessType,
        page: pageIndex,
        resultsPerPage,
        orderBy,
        reverse: reverse.toString(),
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid search query: ' + error.response?.data.message;
    if (status === 401) return 'You have been logged out. Please login again and retry';
    return `Request failed: ${status}`;
  }

  if (!is<SearchResults<Business>>(response.data)) {
    return 'Response is not business array';
  }

  return response.data;
}

/**
 * Makes the provided user an administrator of the provided business.
 *
 * @param businessId Business id to add an administrator to
 * @param userId User id to add to the provided business
 * @returns undefined if operation is successful, otherwise a string error message
 */
export async function makeBusinessAdmin(businessId: number, userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.put(`/businesses/${businessId}/makeAdministrator`, {
      userId,
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'User doesn\'t exist or is already an admin';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return 'Current user cannot perform this action';
    if (status === 406) return 'Business not found';

    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Removes the administrator status of a user from a business
 *
 * @param businessId Business id to remove an administrator from
 * @param userId User id to remove from the provided business
 * @returns undefined if operation is successful, otherwise a string error message
 */
export async function removeBusinessAdmin(businessId: number, userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.put(`/businesses/${businessId}/removeAdministrator`, {
      userId,
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'User doesn\'t exist or is not an admin';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return 'Current user cannot perform this action';
    if (status === 406) return 'The new business admin should be at least 16 years old';

    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Modifies a business given a business id and an updated business object
 * @param businessId Id of the business which is to be updated
 * @param business Business object with all the fields to be updated.
 */
export async function modifyBusiness(businessId: number, business: ModifyBusiness): Promise<MaybeError<undefined>> {
  try {
    await instance.put(`/businesses/${businessId}`, business);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid details entered: ' + error.response?.data.message;
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return 'Invalid authorization for modifying this business';
    if (status === 406) return 'Business does not exist';

    return 'Request failed: ' + error.response?.data.message;
  }
  return undefined;
}