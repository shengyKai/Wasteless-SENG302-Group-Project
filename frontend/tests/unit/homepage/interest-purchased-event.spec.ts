import Vue from "vue";
import Vuetify from "vuetify";
import {createLocalVue, mount, Wrapper} from "@vue/test-utils";
import Vuex from "vuex";
import InterestPurchasedEvent from "@/components/home/newsfeed/InterestPurchasedEvent.vue";
import { Business } from "@/api/business";
import { Product } from "@/api/product";
import { BoughtSale } from "@/api/sale";
import * as events from '@/api/events';

Vue.use(Vuetify);

const business: Business = {
  images:[],
  primaryAdministratorId:2,
  address:{
    country:"Iceland",
    streetName:"Racheal Road",
    streetNumber:"711","city":"London",
    district:"New Plymouth District",
    postcode:"42105",
    region:"Southland"},
  created:"2021-09-14T01:29:16.496459Z",
  name:"Hillary Cresenct Jewelers",
  description:"Nihil, eveniet aliquid culpa officia aut! Impedit sit sunt quaerat, odit, tenetur error, harum nesciunt.",
  id:1,
  businessType:"Accommodation and Food Services"
};
const product: Product = {
  countryOfSale: "Iceland",
  images:[],
  recommendedRetailPrice:842.92,
  business,
  created:"2021-09-14T01:29:16.497456Z",
  name:"Complex Omelette",
  description:"D quam  consequuntur! Commodi minima excepturi repudiandae velit hic maxime doloremque. Quaerat provident commodi consectetur veniam similique ad earum omnis ipsum saepe, volu.",
  id:"FOXLVWMOW603801",
  manufacturer:"Hoffman Incorporated"
};

const boughtSaleItem: BoughtSale = {
  id: 2,
  buyer: null,
  interestCount: 100,
  quantity: 7,
  product,
  price: 50,
  listingDate: "2021-09-14T01:29:16.497456Z",
  saleDate: "2021-09-14T01:29:16.496459Z",
};

const event: events.InterestPurchasedEvent = {
  id: 1,
  type: "InterestPurchasedEvent",
  created: "2021-09-14T01:29:19.731330Z",
  tag: "none",
  status: "normal",
  lastModified: "2021-09-14T01:29:41.214058Z",
  read: true,
  boughtSaleItem,
};

describe("InterestEvent.vue", () => {
  let wrapper: Wrapper<any>;
  let vuetify: Vuetify;

  beforeEach(async () => {
    const localVue = createLocalVue();
    vuetify = new Vuetify();

    localVue.use(Vuex);
    wrapper = mount(InterestPurchasedEvent, {
      localVue,
      vuetify,
      propsData: {
        event,
      },
      stubs: ['router-link'],
    });
  });

  it("The event contains the product name", () => {
    expect(wrapper.text()).toContain("Complex Omelette");
  });

  it("The event contains the business name", () => {
    expect(wrapper.text()).toContain("Hillary Cresenct Jewelers");
  });

  it('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});