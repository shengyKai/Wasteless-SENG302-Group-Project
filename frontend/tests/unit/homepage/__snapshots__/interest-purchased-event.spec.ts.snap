// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterestEvent.vue Matches snapshot 1`] = `
<div>
  <div class="row">
    <div class="col">
      <div aria-haspopup="true" aria-expanded="false" class="v-card__title"><span class="v-badge v-badge--icon v-badge--overlap theme--light">
                An item you have liked has been purchased
              <span class="v-badge__wrapper"><transition-stub name="scale-rotate-transition"><span aria-atomic="true" aria-label="Badge" aria-live="polite" role="status" class="v-badge__badge secondary" style="bottom: calc(100% - 8px); left: calc(100% - -2px);"><i aria-hidden="true" class="v-icon notranslate mdi mdi-email-open theme--light"></i></span></transition-stub></span></span></div><span class="v-tooltip v-tooltip--top"><!----></span>
      <div class="v-card__subtitle">
        14 Sep 2021, 13:29
      </div>
    </div>
    <div class="mt-2 col col-auto"><button type="button" aria-haspopup="true" aria-expanded="false" class="v-icon notranslate mr-2 v-icon--link mdi mdi-star-outline theme--light orange--text"></button><span class="v-tooltip v-tooltip--top"><!----></span> <span class="v-tooltip v-tooltip--top"><!----></span><button type="button" aria-haspopup="true" aria-expanded="false" class="v-icon notranslate mr-2 v-icon--link mdi mdi-archive-outline theme--light blue--text"></button> <span class="v-tooltip v-tooltip--top"><!----></span><button type="button" aria-haspopup="true" aria-expanded="false" class="v-icon notranslate mr-2 v-icon--link mdi mdi-trash-can theme--light red--text"></button></div>
  </div>
  <div class="v-card__text pb-1"><strong>7x Complex Omelette</strong>
    from
    <strong><router-link-stub to="/business/1" class="text--secondary">
        Hillary Cresenct Jewelers
      </router-link-stub></strong>
    has been sold to another user and is no longer available for purchase.
  </div>
  <!---->
  <div class="row">
    <div class="col"><span class="ml-4 v-chip v-chip--clickable v-chip--label theme--light v-size--default none white--text" aria-haspopup="true" aria-expanded="false"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i>
            Tag
          </span></span><span class="v-tooltip v-tooltip--bottom"><!----></span></div>
  </div>
  <div class="row no-gutters">
    <div class="col">
      <transition-stub name="expand-transition">
        <div class="mx-auto v-card v-sheet theme--light" style="max-width: 1200px; display: none;">
          <div class="font-weight-medium"><span class="ml-4">
              Change your Tag:
            </span></div>
          <div class="ml-3"><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default none white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default red white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default orange white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default yellow white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default green white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default blue white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span><span class="ma-1 v-chip v-chip--clickable v-chip--label theme--light v-size--default purple white--text"><span class="v-chip__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-label theme--light"></i></span></span></div>
        </div>
      </transition-stub>
    </div>
  </div>
</div>
`;
